controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -321,6 +322,9 @@ public void unregisterBucketOwnershipListener() {
     public void registerBucketChangeListener(int bucket, BucketChangeListener listener) {
         Preconditions.checkNotNull(listener);
 
+        String bucketRoot = String.format(BUCKET_PATH, bucket);
```

Comment:
This change does not seem right. There are other listeners in this class too, like the ownership one. Also, this is the wrong place to do this. If we are to create this bucket znode, it should be during the process of acquiring ownership, not while registering a listener. 

I need to do some further investigation before making a concrete proposal here.

Reply:
Follow-up question: why not create all necessary persistent znodes during the initialization of the controller process?

https://github.com/pravega/pravega/blob/master/controller/src/main/java/io/pravega/controller/server/Main.java

Reply:
@fpj This is another possibility that may be interesting, given that I observed the same phenomenon related to other znodes. This can be done when instantiating a new `ZKStreamMetadataStore` object. In fact, `ZKHostStore` does something similar in the method `tryInit()`, but instead of ensuring that the znode is created in the constructor, it is check on every call to Zookeeper. I will try to find a better, broader znode initialization approach.

Reply:
Following your suggestion, I added a new method to `StreamMetadataStore` interface called `initializeMetadataStore()`. In the case of `ZKStreamMetadataStore`, it contains the logic to initialize the Stream buckets. This method is invoked in `ControllerServiceStarter`, just after the `StreamMetadataStore` object gets created. This may be a suitable "centralized" point to initialize all the necessary metadata items to facilitate maintenance.

Note that this solution fixes the reported problem (system tests are passing with the current version of the code) and could be easily cherry-picked to previous versions. But it could subject to changes in PR #3230, given that the initialization of "buckets" will be done for multiple services, not just for the retention service.